### ISSUE TITLE
Clarification on feature grouping training when final feature counts differ after delete_unique_features

### ISSUE THREAD

**Issue by jamesYu365** [soda-inria/tabicl#131]
Hi, thank you again for open-sourcing TabICLv2 and the PR #111 graph_scm prior.

I am trying to reproduce a smaller-scale TabICLv2 classification pretraining run using PR #111, with the v2 feature grouping path enabled (`col_feature_group=same`).

While wiring the training loop, I noticed a potential ambiguity around variable final feature counts (`d`) after prior generation.

From the current code:

- `ColEmbedding.forward` asserts that `d is None` when feature grouping is enabled:
  `assert d is None, "d is not supported when feature grouping is enabled."`
- the trainer aligns each micro-batch by truncating `micro_X` to `micro_d.max()`, then calls:
  `self.model(micro_X, y_train, micro_d)`
- `TabICL._train_forward` only sets `d=None` when all `d` values are identical and equal to the current feature dimension `H`.

I initially thought setting `batch_size_per_gp` / `batch_size_per_subgp` equal to the per-rank micro-batch size would solve this, since the prior samples one shared `num_features` per subgroup. However, `delete_unique_features` removes constant features independently per generated dataset and updates the final `d`.

In a small prior-only diagnostic with the Stage-1-like setting:

- global batch size: 66
- simulated world size: 6
- per-rank batch / micro-batch size: 11
- `batch_size_per_gp=11`
- `batch_size_per_subgp=11`
- `col_feature_group=same`
- `graph_scm` prior with filtering enabled

I found that even though the sampled `num_features` is shared within the micro-batch, the final `d` after `delete_unique_features` often differs across datasets.

For example:

```text
[55, 44, 55, 55, 55, 53, 55, 55, 55, 53, 53]
[49, 47, 50, 50, 50, 49, 50, 50, 50, 50, 50]
[38, 45, 32, 50, 50, 34, 47, 50, 36, 50, 50]
```

In my small diagnostic:

- with `min_features=5,max_features=100`, 46/50 micro-batches had mixed final `d`;
- with fixed `min_features=max_features=50`, 49/50 micro-batches still had mixed final `d`.

So I wanted to ask:

1. In the paper-aligned TabICLv2 pretraining recipe with feature grouping enabled, how do you handle this final `d` mismatch after `delete_unique_features`?
2. Is the intended solution to use `micro_batch_size=1` / per-dataset forward with gradient accumulation?
3. Or do you bucket/regroup generated datasets by final `d` before the model forward?

This matters because the public PR111 trainer seems to work only when each feature-grouping forward receives `d=None`, but the current prior can produce mixed final `d` within a micro-batch even when sampled `num_features` is shared.

Thank you!

**Comment by dholzmueller** [comment]
1. We just ignore d in the architecture (I guess you could remove the assert).
2. We use micro_batch_size=4 for stage 1 of pre-training and =1 for the other two stages.
3. see 1.

**Comment by jamesYu365** [comment]
Thanks, that makes sense. So for TabICLv2 with feature grouping, my understanding is that `d` is simply not used in the architecture: the batch is cropped to the maximum feature count in the micro-batch, and any zero-padded columns are treated as regular zero-valued inputs rather than being masked. Is that correct?

**Comment by dholzmueller** [comment]
Yes, that is correct.

**Comment by jamesYu365** [comment]
Thanks, that confirms the implementation detail.
